### PR TITLE
[DataPipe] Preventing automatic reset call after state is restored

### PR DESCRIPTION
**Pull request by NivekT**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #70216
* #70373
* #77775
* __->__ #77774

Fixes https://github.com/pytorch/data/issues/419

Differential Revision: [D36521039](https://our.internmc.facebook.com/intern/diff/D36521039)

### CODE REVIEW

**Review by ejguan — COMMENTED** [review]
We are currently automatically invoke `reset()` for `__iter__`. Could we wrap `self.reset()` by this logic in `_typing.py`? Then, users won't need to add this duplicate logic everywhere.

**Comment by NivekT** [comment]
@ejguan I have refactored to modify `IterDataPipe`'s `reset` and `__setstate__` in `_typing.py` instead. If it looks good, you can remove `_restored` from `Shuffler`.

File: torch/utils/data/datapipes/_typing.py [hunk 1]
```diff
@@ -339,10 +339,44 @@ def _hash_(self):
...
 class _IterDataPipeMeta(_DataPipeMeta):
     r"""
-    Metaclass for `IterDataPipe` and inherits from `_DataPipeMeta`. Aad a hook function to `__iter__`.
+    Metaclass for `IterDataPipe` and inherits from `_DataPipeMeta`. Aad various functions for behaviors
+    specific to `IterDataPipe`.
     """
 
     def __new__(cls, name, bases, namespace, **kwargs):
+
+        if 'reset' in namespace:
```
Comment (ejguan) [inline torch/utils/data/datapipes/_typing.py:348]: I think `reset` is always in the namespace.

**Review by ejguan — APPROVED** [review]
LGTM with a few comments below

**Comment by janeyx99** [comment]
Ah, sorry @NivekT, my builds required change got reverted so merge -g didn't check for them anymore. Also we realize merge -g is a bit of a misnomer since it only waits for required checks to pass vs all checks and are fixing that too so that merge -g will mean all checks in the future.

File: torch/utils/data/datapipes/_typing.py (hunk 1 above)
Comment (NivekT) [reply torch/utils/data/datapipes/_typing.py]: Actually it is not always in the namespace. It is only in the namespace if the class has the method `reset` defined, it doesn't seem like it is inheriting from `IterDataPipe`'s namespace at this point.

Comment (ejguan) [reply torch/utils/data/datapipes/_typing.py]: #TIL

**Comment by NivekT** [comment]
> > @pytorchbot merge this please -g
> 
> @NivekT just FYI I'm not sure this actually does the -g functionality. cc: @zengk95

I was expecting it would since you re-landed your change. Is that separate?

In either case, I think this PR should be okay.

**Comment by NivekT** [comment]
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36521039).

**Comment by janeyx99** [comment]
Ok yea this PR is okay, I just mean that in the future, it's just `@pytorchbot merge -g`

**Pull request closed without merging**